From 31dfc909f180d4e91422fd038fd6d3c4228c3aa0 Mon Sep 17 00:00:00 2001
From: Peter Hutterer <peter.hutterer@who-t.net>
Date: Wed, 28 Jul 2010 10:10:26 +1000
Subject: [PATCH] Gesture: send ctrl + scroll wheel events instead of +/-.

The number of apps supporting scroll wheel zoom events is higher than the
ones supporting ctrl +/-. And the ctrl key can be assumed on the same key
for most keyboards, leaving us with a more stable zoom gesture than the
previous approach of hardcoding the position of +/-.

Suggested-by: Ping Cheng <pinglinux@gmail.com>
Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
Acked-by: Ping Cheng <pinglinux@gmail.com>

Conflicts:

	src/wcmTouchFilter.c
---
 src/wcmTouchFilter.c |   31 ++++++++-----------------------
 1 files changed, 8 insertions(+), 23 deletions(-)

diff --git a/src/wcmTouchFilter.c b/src/wcmTouchFilter.c
index e40ea2e..dc23a61 100644
--- a/src/wcmTouchFilter.c
+++ b/src/wcmTouchFilter.c
@@ -397,29 +397,14 @@ static void wcmFingerZoom(WacomDevicePtr priv)
 	/* zooming? */
 	if (abs(dist) > WACOM_MOTION_IN_POINT)
 	{
-		for (i=0; i<(int)(((double)abs(dist)/
-				(double)WACOM_MOTION_IN_POINT) + 0.5); i++)
-		{
-			/* FIXME: this hardcodes the positions of ctrl, + and - to the ones on
-			   the us keyboard layout. Tough luck. The alternative is to run
-			   through the XKB table and figure out where +/- are hiding. Good
-			   luck. Gesture support is not supposed to be in the driver...
-			 */
-			int key = (dist > 0) ? 21 /*XK_plus*/ : 20 /*XK_minus*/;
-			wcmEmitKeycode (priv->local->dev, 37 /*XK_Control_L*/, 1);
-			/* zooming in */
-			if (dist > 0)
-			{
-				wcmEmitKeycode (priv->local->dev, key, 1);
-				wcmEmitKeycode (priv->local->dev, key, 0);
-			}
-			else /* zooming out */
-			{
-				wcmEmitKeycode (priv->local->dev, key, 1);
-				wcmEmitKeycode (priv->local->dev, key, 0);
-			}
-			wcmEmitKeycode (priv->local->dev, 37 /*XK_Control_L*/, 0);
-		}
+		/* FIXME: this hardcodes the positions of ctrl to the ones on
+		   the us keyboard layout. Tough luck. The alternative is to run
+		   through the XKB table and figure out where it's hiding. Good
+		   luck. Gesture support is not supposed to be in the driver...
+		 */
+		wcmEmitKeycode (priv->local->dev, 37 /*XK_Control_L*/, 1);
+		wcmSendVerticalScrollEvent(priv, dist, 4, 5);
+		wcmEmitKeycode (priv->local->dev, 37 /*XK_Control_L*/, 0);
 
 		/* reset initial states */
 		common->wcmGestureState[0] = ds[0];
-- 
1.7.1

